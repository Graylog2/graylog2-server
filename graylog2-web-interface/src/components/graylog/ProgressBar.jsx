// eslint-disable-next-line no-restricted-imports
import { ProgressBar } from 'react-bootstrap';
import { css } from 'styled-components';
import { transparentize } from 'polished';

import teinte from 'theme/teinte';
import { variantColors } from './variants/bsStyle';

const defaultStripColor = transparentize(0.75, teinte.primary.due);

const variants = (styles) => {
  let style = '';

  styles.forEach((variant) => {
    style += `.progress-bar-${variant} {
      background-color: ${variantColors[variant]};
    }`;
  });

  return css`${style}`;
};

export const progressBarStyles = css`
  .progress {
    background-color: ${teinte.secondary.due};

    .progress-bar {
      color: ${teinte.primary.due};
      background-color: ${teinte.tertiary.uno};
    }

    .progress-striped .progress-bar,
    .progress-bar-striped {
      background-image: linear-gradient(45deg,
                        ${defaultStripColor} 25%,
                        transparent 25%,
                        transparent 50%,
                        ${defaultStripColor} 50%,
                        ${defaultStripColor} 75%,
                        transparent 75%,
                        transparent);
    }

    ${variants(['success', 'info', 'warning', 'danger'])};
  }
`;

export default ProgressBar;
