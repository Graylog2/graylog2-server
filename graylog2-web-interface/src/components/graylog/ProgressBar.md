## Default
```js
<ProgressBar bars={[{ value: 35 }]} />
```

## Labeled
```js
const value = 60;
<ProgressBar bars={[{ value, label: `${value}%` }]} />
```

## Variant
```js
<div>
  <ProgressBar bars={[{ value: 70, bsStyle: 'primary' }]} />
  <ProgressBar bars={[{ value: 40, bsStyle: 'success' }]} />
  <ProgressBar bars={[{ value: 20, bsStyle: 'info' }]} />
  <ProgressBar bars={[{ value: 60, bsStyle: 'warning' }]} />
  <ProgressBar bars={[{ value: 80, bsStyle: 'danger' }]} />
</div>
```

## Striped
```js
<div>
  <ProgressBar bars={[{ value: 40, bsStyle: 'success', striped: true }]} />
  <ProgressBar bars={[{ value: 20, bsStyle: 'info', striped: true }]} />
  <ProgressBar bars={[{ value: 60, bsStyle: 'warning', striped: true }]} />
  <ProgressBar bars={[{ value: 80, bsStyle: 'danger', striped: true }]} />
</div>
```
## Animated

```js
  <ProgressBar bars={[{ value: 45, bsStyle: 'danger', animated: true }]} />
```

## Multiples
```js
<ProgressBar bars={[
  { value: 35, bsStyle: 'success', striped: true },
  { value: 20, bsStyle: 'warning' },
  { value: 10, bsStyle: 'danger', animated: true }
]} />
```
