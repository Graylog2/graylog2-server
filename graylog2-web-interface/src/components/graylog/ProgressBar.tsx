/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import styled, { css, keyframes } from 'styled-components';
import chroma from 'chroma-js';

type StyledBarProps = {
  animated: boolean,
  bsStyle: string,
  striped: boolean,
  value: number,
};

type BarProps = StyledBarProps & {
  label: string,
};

type ProgressBarProps = {
  bars: Array<BarProps>,
  className: string,
};

const DEFAULT_BAR = {
  animated: false,
  bsStyle: 'info',
  label: undefined,
  striped: false,
  value: 0,
};

const boxShadow = (meta) => css`
  box-shadow: ${meta} ${({ theme }) => chroma(theme.colors.brand.secondary).alpha(0.1).css()};
`;

const animatedStripes = keyframes`
  from {
    background-position: 40px 0;
  }

  to {
    background-position: 0 0;
  }
`;

const progressBarVariants = css<StyledBarProps>(({ bsStyle, theme }) => {
  if (!bsStyle) {
    return undefined;
  }

  return `
    background-color: ${theme.colors.variant[bsStyle]};
    color: ${theme.utils.readableColor(theme.colors.variant[bsStyle])};
  `;
});

const ProgressWrap = styled.div(({ theme }) => css`
  height: 20px;
  margin-bottom: 20px;
  overflow: hidden;
  background-color: ${theme.colors.gray[90]};
  border-radius: 4px;
  ${boxShadow('inset 0 1px 2px')}
  display: flex;
  align-items: center;
`);

const Bar = styled.div<StyledBarProps>(({ animated, striped, theme, value }) => {
  const defaultStripColor = chroma(theme.colors.global.contentBackground).alpha(0.25).css();

  return css`
    height: 100%;
    font-size: ${theme.fonts.size.small};
    line-height: 20px;
    text-align: center;
    ${boxShadow('inset 0 -1px 0')}
    transition: width 500ms ease-in-out;
    width: ${value}%;
    max-width: 100%;
    text-shadow: 0 1px 2px ${chroma(theme.colors.gray[10]).alpha(0.4).css()}, 2px -1px 3px ${chroma(theme.colors.gray[100]).alpha(0.5).css()};
    ${(animated || striped) && css`
      background-image: linear-gradient(
        45deg,
        ${defaultStripColor} 25%,
        transparent 25%,
        transparent 50%,
        ${defaultStripColor} 50%,
        ${defaultStripColor} 75%,
        transparent 75%,
        transparent
      );
      background-size: 40px 40px;
    `}
    ${animated && css`
      animation: ${animatedStripes} 2s linear infinite;
    `}
    ${progressBarVariants}
  `;
});

const ProgressBar = ({ bars, className }: ProgressBarProps) => {
  return (
    <ProgressWrap className={className}>
      {bars.map((bar, index) => {
        const { label, animated, bsStyle, striped, value } = { ...DEFAULT_BAR, ...bar };

        return (
          <Bar role="progressbar"
               aria-valuenow={value}
               aria-valuemin={0}
               aria-valuemax={100}
               aria-valuetext={label}
               key={`bar-${index}`} // eslint-disable-line react/no-array-index-key
               animated={animated}
               bsStyle={bsStyle}
               striped={striped}
               value={value}>
            {label}
          </Bar>
        );
      })}
    </ProgressWrap>
  );
};

ProgressBar.propTypes = {
  bars: PropTypes.arrayOf(PropTypes.shape({
    animated: PropTypes.bool,
    bsStyle: PropTypes.string,
    label: PropTypes.string,
    striped: PropTypes.bool,
    value: PropTypes.number,
  })),
  className: PropTypes.string,
};

ProgressBar.defaultProps = {
  bars: [DEFAULT_BAR],
  className: undefined,
};

export default ProgressBar;
export { Bar };
