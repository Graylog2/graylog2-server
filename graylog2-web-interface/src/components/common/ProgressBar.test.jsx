/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import React from 'react';
import { render } from 'wrappedTestingLibrary';

import ProgressBar from './ProgressBar';

describe('<ProgressBar />', () => {
  it('properly renders with no props', () => {
    const { container } = render(<ProgressBar />);

    expect(container).toMatchSnapshot();
  });

  it('properly renders with one bar', () => {
    const { container } = render(<ProgressBar bars={[{ value: 35 }]} />);

    expect(container).toMatchSnapshot();
  });

  it('properly renders with multiple bars', () => {
    const { container } = render(<ProgressBar bars={[{ value: 35 }, { value: 55 }]} />);

    expect(container).toMatchSnapshot();
  });

  it('properly renders with label', () => {
    const { container } = render(<ProgressBar bars={[{ value: 23, label: 'Example ProgressBar' }]} />);

    expect(container).toMatchSnapshot();
  });

  it('properly renders with animated', () => {
    const { container } = render(<ProgressBar bars={[{ value: 45, animated: true }]} />);

    expect(container).toMatchSnapshot();
  });

  it('properly renders with striped', () => {
    const { container } = render(<ProgressBar bars={[{ value: 56, striped: true }]} />);

    expect(container).toMatchSnapshot();
  });

  it('properly renders with bsStyle variant', () => {
    const { container } = render(<ProgressBar bars={[{ value: 67, bsStyle: 'danger' }]} />);

    expect(container).toMatchSnapshot();
  });
});
