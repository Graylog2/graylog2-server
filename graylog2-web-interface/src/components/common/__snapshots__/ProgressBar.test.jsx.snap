// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> properly renders with animated 1`] = `
.c0 {
  height: 20px;
  margin-bottom: 20px;
  overflow: hidden;
  background-color: #eaeaeb;
  border-radius: 4px;
  box-shadow: inset 0 1px 2px rgba(255,255,255,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  height: 100%;
  font-size: 0.878rem;
  text-align: center;
  box-shadow: inset 0 -1px 0 rgba(255,255,255,0.1);
  -webkit-transition: width 500ms ease-in-out;
  transition: width 500ms ease-in-out;
  width: 45%;
  max-width: 100%;
  text-shadow: 0 1px 2px rgba(62,67,76,0.4),2px -1px 3px rgba(255,255,255,0.5);
  background-image: linear-gradient( 45deg,rgba(255,255,255,0.25) 25%,transparent 25%,transparent 50%,rgba(255,255,255,0.25) 50%,rgba(255,255,255,0.25) 75%,transparent 75%,transparent );
  background-size: 40px 40px;
  -webkit-animation: jWjlvG 2s linear infinite;
  animation: jWjlvG 2s linear infinite;
  background-color: #578dcc;
  color: #3e434c;
}

<div>
  <div
    class="c0"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="45"
      class="c1"
      role="progressbar"
      value="45"
    />
  </div>
</div>
`;

exports[`<ProgressBar /> properly renders with bsStyle variant 1`] = `
.c0 {
  height: 20px;
  margin-bottom: 20px;
  overflow: hidden;
  background-color: #eaeaeb;
  border-radius: 4px;
  box-shadow: inset 0 1px 2px rgba(255,255,255,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  height: 100%;
  font-size: 0.878rem;
  text-align: center;
  box-shadow: inset 0 -1px 0 rgba(255,255,255,0.1);
  -webkit-transition: width 500ms ease-in-out;
  transition: width 500ms ease-in-out;
  width: 67%;
  max-width: 100%;
  text-shadow: 0 1px 2px rgba(62,67,76,0.4),2px -1px 3px rgba(255,255,255,0.5);
  background-color: #eb5454;
  color: #3e434c;
}

<div>
  <div
    class="c0"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="67"
      class="c1"
      role="progressbar"
      value="67"
    />
  </div>
</div>
`;

exports[`<ProgressBar /> properly renders with label 1`] = `
.c0 {
  height: 20px;
  margin-bottom: 20px;
  overflow: hidden;
  background-color: #eaeaeb;
  border-radius: 4px;
  box-shadow: inset 0 1px 2px rgba(255,255,255,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  height: 100%;
  font-size: 0.878rem;
  text-align: center;
  box-shadow: inset 0 -1px 0 rgba(255,255,255,0.1);
  -webkit-transition: width 500ms ease-in-out;
  transition: width 500ms ease-in-out;
  width: 23%;
  max-width: 100%;
  text-shadow: 0 1px 2px rgba(62,67,76,0.4),2px -1px 3px rgba(255,255,255,0.5);
  background-color: #578dcc;
  color: #3e434c;
}

<div>
  <div
    class="c0"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="23"
      aria-valuetext="Example ProgressBar"
      class="c1"
      role="progressbar"
      value="23"
    >
      Example ProgressBar
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> properly renders with multiple bars 1`] = `
.c0 {
  height: 20px;
  margin-bottom: 20px;
  overflow: hidden;
  background-color: #eaeaeb;
  border-radius: 4px;
  box-shadow: inset 0 1px 2px rgba(255,255,255,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  height: 100%;
  font-size: 0.878rem;
  text-align: center;
  box-shadow: inset 0 -1px 0 rgba(255,255,255,0.1);
  -webkit-transition: width 500ms ease-in-out;
  transition: width 500ms ease-in-out;
  width: 35%;
  max-width: 100%;
  text-shadow: 0 1px 2px rgba(62,67,76,0.4),2px -1px 3px rgba(255,255,255,0.5);
  background-color: #578dcc;
  color: #3e434c;
}

.c2 {
  height: 100%;
  font-size: 0.878rem;
  text-align: center;
  box-shadow: inset 0 -1px 0 rgba(255,255,255,0.1);
  -webkit-transition: width 500ms ease-in-out;
  transition: width 500ms ease-in-out;
  width: 55%;
  max-width: 100%;
  text-shadow: 0 1px 2px rgba(62,67,76,0.4),2px -1px 3px rgba(255,255,255,0.5);
  background-color: #578dcc;
  color: #3e434c;
}

<div>
  <div
    class="c0"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="35"
      class="c1"
      role="progressbar"
      value="35"
    />
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="55"
      class="c2"
      role="progressbar"
      value="55"
    />
  </div>
</div>
`;

exports[`<ProgressBar /> properly renders with no props 1`] = `
.c0 {
  height: 20px;
  margin-bottom: 20px;
  overflow: hidden;
  background-color: #eaeaeb;
  border-radius: 4px;
  box-shadow: inset 0 1px 2px rgba(255,255,255,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  height: 100%;
  font-size: 0.878rem;
  text-align: center;
  box-shadow: inset 0 -1px 0 rgba(255,255,255,0.1);
  -webkit-transition: width 500ms ease-in-out;
  transition: width 500ms ease-in-out;
  width: 0%;
  max-width: 100%;
  text-shadow: 0 1px 2px rgba(62,67,76,0.4),2px -1px 3px rgba(255,255,255,0.5);
  background-color: #578dcc;
  color: #3e434c;
}

<div>
  <div
    class="c0"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      class="c1"
      role="progressbar"
      value="0"
    />
  </div>
</div>
`;

exports[`<ProgressBar /> properly renders with one bar 1`] = `
.c0 {
  height: 20px;
  margin-bottom: 20px;
  overflow: hidden;
  background-color: #eaeaeb;
  border-radius: 4px;
  box-shadow: inset 0 1px 2px rgba(255,255,255,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  height: 100%;
  font-size: 0.878rem;
  text-align: center;
  box-shadow: inset 0 -1px 0 rgba(255,255,255,0.1);
  -webkit-transition: width 500ms ease-in-out;
  transition: width 500ms ease-in-out;
  width: 35%;
  max-width: 100%;
  text-shadow: 0 1px 2px rgba(62,67,76,0.4),2px -1px 3px rgba(255,255,255,0.5);
  background-color: #578dcc;
  color: #3e434c;
}

<div>
  <div
    class="c0"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="35"
      class="c1"
      role="progressbar"
      value="35"
    />
  </div>
</div>
`;

exports[`<ProgressBar /> properly renders with striped 1`] = `
.c0 {
  height: 20px;
  margin-bottom: 20px;
  overflow: hidden;
  background-color: #eaeaeb;
  border-radius: 4px;
  box-shadow: inset 0 1px 2px rgba(255,255,255,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  height: 100%;
  font-size: 0.878rem;
  text-align: center;
  box-shadow: inset 0 -1px 0 rgba(255,255,255,0.1);
  -webkit-transition: width 500ms ease-in-out;
  transition: width 500ms ease-in-out;
  width: 56%;
  max-width: 100%;
  text-shadow: 0 1px 2px rgba(62,67,76,0.4),2px -1px 3px rgba(255,255,255,0.5);
  background-image: linear-gradient( 45deg,rgba(255,255,255,0.25) 25%,transparent 25%,transparent 50%,rgba(255,255,255,0.25) 50%,rgba(255,255,255,0.25) 75%,transparent 75%,transparent );
  background-size: 40px 40px;
  background-color: #578dcc;
  color: #3e434c;
}

<div>
  <div
    class="c0"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="56"
      class="c1"
      role="progressbar"
      value="56"
    />
  </div>
</div>
`;
